README.md
PyBal is a LVS monitor. It monitors HTTP or DNS servers and adapts LVS state
based on the results. PyBal was created for use by Wikimedia.

[![Build Status](https://travis-ci.org/wikimedia/PyBal.svg?branch=master)](https://travis-ci.org/wikimedia/PyBal)
[![Coverage Status](https://img.shields.io/coveralls/wikimedia/PyBal.svg)](https://coveralls.io/r/wikimedia/PyBal?branch=master)

Unit tests are available under pybal/test/. To run the unit tests and get a
coverage report:

    pip install -r requirements.txt
    coverage run --source=pybal setup.py test && coverage report

Branching/release policy
-------------

Since PyBal is a mission-critical software and a bug can bring down
the whole infrastructure, we try to minimize the risk by avoiding
having to release large changes when we want to bring a bugfix to
production. In order to do so, we do the following:

- We adopt [semantic versioning 2.0 ](http://semver.org/) so we change
  the major version for API breaking changes, the minor version for
  adding new functionality, and the patch version for bugfixes.
- All development of non breaking new features should happen against
  the master branch, with the exception of minor version-specific
  bugfixes
- Breaking changes (ones that would trigger a major revision change)
  should first be developed in a separate development branch named
  X.0-dev. It will be merged back into master once we're ready to move
  to a new major release. This way we can work on/test the breaking
  change while not stopping further development of incremental
  improvements on the current codebase.
- Whenever we're happy with the new functionality in master and we
  think we're ready to take that to production, we create a branch
  named as the minor version (e.g. "1.09"). We'll call these the
  "release branches".
- When we make a bugfix that applies to the code in master as well, we
  do the CR/merge on the master branch, and then we cherry-pick the
  change to the release branches currently maintained.
  If the bugfix is specific to a release branch (because, for example,
  the code in master has been rewritten/replaced), we just post it
  against the release branch.
- New versions should always come from release branches.
- When we decide we don't actively maintain a minor version, a commit
  to the release branch should be done indicating that.

So let's review how a typical bugfix workflow works:

- The patch to fix the bug is developed and reviewed on master.
- Once it's merged on master, the patch is cherry-picked to the
  release branches and either merged (if the cherry-pick is clean) or
  re-reviewed: please apply common sense
- When we're happy with the patch, we create the debian/changelog
  entry on the release branch. This entry will need to be reintegrated
  back to the debian/changelog on master, most probably manually in a
  separate commit.

If your commit is instead adding a new functionality, it should definitely go
to master. If it's introducing breaking changes, it goes to the
X.0-dev branch currently under development.


debian/changelog
pybal (1.15.9) buster-wikimedia; urgency=medium

  * Reduce reconnect timeout for etcd to 0.1 seconds (T264362)

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Wed, 27 Jan 2021 12:46:38 +0100

pybal (1.15.8) buster-wikimedia; urgency=medium

  * Build for buster (T245984)

 -- Valentin Gutierrez <vgutierrez@wikimedia.org>  Mon, 24 Feb 2020 11:06:04 +0000

pybal (1.15.7) stretch-wikimedia; urgency=medium

  * Handle etcd 401 errors

 -- Valentin Gutierrez <vgutierrez@wikimedia.org>  Tue, 21 Jan 2020 11:03:12 +0000

pybal (1.15.6) stretch-wikimedia; urgency=medium
  * Fix proxy fetch multi URL cherry-pick

 -- Valentin Gutierrez <vgutierrez@wikimedia.org>  Tue, 07 May 2019 17:32:34 +0200

pybal (1.15.5) stretch-wikimedia; urgency=medium
  * Rollback kubernetes configuration observer due to a dependency issue on jessie

 -- Valentin Gutierrez <vgutierrez@wikimedia.org>  Tue, 07 May 2019 16:35:22 +0200

pybal (1.15.4) stretch-wikimedia; urgency=medium
  * Allow proxyfetch to check more than one url at a time
  * Avoid Deferred.cancel() induced CancelledErrors
  * Add unit tests for ProxyFetchMonitoringProtocol
  * Add minimal test cases for Skeleton and ProxyFetch
  * Create MonitoringProtocolTestCase base class
  * Implement kubernetes configuration observer (Bug: T192437)

 -- Valentin Gutierrez <vgutierrez@wikimedia.org>  Tue, 07 May 2019 15:54:19 +0200

pybal (1.15.3) stretch-wikimedia; urgency=medium
  * First Debian stretch release (Bug: T177961)
  * Fix dummy metrics implementation (Bug: T190527)
    49d297585edfb5fd9152cd6675904bd06656fbbd
  * Bump python package version to 1.15
    d2535baff3aae405232959db359cbcf46450a8e6

 -- Valentin Gutierrez <vgutierrez@wikimedia.org>  Tue, 03 Apr 2018 15:49:49 +0200

pybal (1.15.2) jessie-wikimedia; urgency=medium

  * Ensure per-service-MED is an integer
    c22dae5bbffe6dd2faf1a254f3992f3dad59987c

 -- Valentin Gutierrez <vgutierrez@wikimedia.org>  Thu, 08 Mar 2018 14:19:17 +0100

pybal (1.15.1) jessie-wikimedia; urgency=medium

  * Fix MPReachNLRIAttribute generation
    77430defa915ad09310a109e0acc6aac0f557635

 -- Valentin Gutierrez <vgutierrez@wikimedia.org>  Tue, 06 Mar 2018 17:58:44 +0100

pybal (1.15.0) jessie-wikimedia; urgency=medium

  * Add peer address information to BGP and FSM log lines (Bug: T188085)
    8c2dce2d9c4400bbba137de2554adcbcfe6576f3

  * Fix inconsistent label name (Bug: T188085)
    c150bd3b8a5d472e951328f6337e99d5a936ac8a

  * Fix log method invocation (Bug: T188085)
    c58c7d283ac373911b5d8f5f69073956f19ffa4c

  * Prefix log lines with the class name (Bug: T188085)
    f3161bf1e86baf9a6b809c01bd218640023d25ee

  * Improve naming of the new BGP metric names and labels (Bug: T188085)
    4f787286bca01ce55817cd304b7087328cc09513

  * Provide unique logging for BGP instances (Bug: T188085)
    8181b72447138d3eeafd76dae515fb3833b3870e

  * Provide an UDP monitor (Bug: T178151)
    d3358e9d7e36cd9b7872bd94b83864ac271bd804

  * Add a test case for removing previously existing servers
    45ead115b03fcb098ea9f568b44e95e09be055fe

  * Provide testing for FSM.BGPTimer (Bug: T188085)
    97219b8f3fa9ed6147ca428fcc989be64018a564

  * Add unit tests for Coordinator methods
    9dc84cc5f9dc8b4a3a6a87ee3b9f0b4d5bec4937

  * Provide BGP session state visibility for every ASN/peer (Bug: T188085)
    bbe4f1f3f95eb457b88cd000a1a81170169d8523

  * Provide test cases for BGP parsing (Bug: T188085)
    b82f87a0ef390aee3f11524dfe5c881b3e693c9d

  * etcd: Improve reactor mocking (Bug: T169765)
    309f08ade5fcc9c701d34fb5c2b82ba6f029521e

 -- Valentin Gutierrez <vgutierrez@wikimedia.org>  Mon, 05 Mar 2018 12:33:06 +0100

pybal (1.14.4) jessie-wikimedia; urgency=medium

  * etcd: Introduce reconnectTimeout (Bug: T169765)
    c917f59f4c37c7b89f15d0fd8b375de028d4a90b

 -- Emanuele Rocca <ema@wikimedia.org>  Wed, 21 Feb 2018 12:40:43 +0100

pybal (1.14.3) jessie-wikimedia; urgency=medium

  * Use up-and-enabled servers in can-depool logic (Bug: T184715)
    1264a784bbcbedee37466102246775e2ee26367a

  * Alerts instrumentation: return instance of bytes (Bug: T184721)
    3505626c9d1590e84525560edbc2f5bf01a7be99

 -- Emanuele Rocca <ema@wikimedia.org>  Wed, 17 Jan 2018 16:49:48 +0100

pybal (1.14.2) jessie-wikimedia; urgency=medium


pybal/bgp/fsm.py
# bgp.fsm.py
# Copyright (c) 2007-2018 by Mark Bergsma <mark@nedworks.org>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
An implementation of the BGP FSM as defined by RFC 4271
"""

# System imports
import logging

# Twisted imports
from twisted.internet import reactor, error

# BGP imports
from constants import *
from exceptions import NotificationSent

# Pybal imports :-(
from pybal.metrics import Gauge
from pybal.util import _log


class FSM(object):
    class BGPTimer(object):
        """
        Timer class with a slightly different Timer interface than the
        Twisted DelayedCall interface
        """

        def __init__(self, callable):
            self.delayedCall = None
            self.callable = callable

        def cancel(self):
            """Cancels the timer if it was running, does nothing otherwise"""

            try:
                self.delayedCall.cancel()
            except (AttributeError, error.AlreadyCalled, error.AlreadyCancelled):
                pass

        def reset(self, secondsFromNow):
            """Resets an already running timer, or starts it if it wasn't running."""

            try:
                self.delayedCall.reset(secondsFromNow)
            except (AttributeError, error.AlreadyCalled, error.AlreadyCancelled):
                self.delayedCall = reactor.callLater(secondsFromNow, self.callable)

        def active(self):
            """Returns True if the timer was running, False otherwise."""

            try:
                return self.delayedCall.active()
            except AttributeError:
                return False

    protocol = None

    state = ST_IDLE

    largeHoldTime = 4*60
    sendNotificationWithoutOpen = True    # No bullshit

    eventMethods = {
        1:  'manualStart',
        2:  'manualStop',
        3:  'automaticStart',
        4:  'manualStartPassive',
        5:  'automaticStartPassive',

        9:  'connectRetryTimeEvent',
        10: 'holdTimeEvent',
        11: 'keepAliveEvent',
        12: 'delayOpenEvent',
        13: 'idleHoldTimeEvent',
        14: 'tcpConnectionValid',

        17: 'connectionMade',
        18: 'connectionFailed',
        19: 'openReceived',
        20: 'openReceived',
        21: 'headerError',
        22: 'openMessageError',
        23: 'openCollisionDump',
        24: 'versionError',
        25: 'notificationReceived',
        26: 'keepAliveReceived',
        27: 'updateReceived',
        28: 'updateError'
    }

    bgpTimers = { 'connectRetryTimer', 'holdTimer', 'keepAliveTimer',
        'delayOpenTimer', 'idleHoldTimer' }

    metric_labelnames = {'local_asn', 'state', 'local_ip', 'remote_ip', 'side'}
    metric_keywords = {
        'labelnames': metric_labelnames,
        'namespace': 'pybal',
        'subsystem': 'bgp'
    }

    metrics = {
        'bgp_session_state_count': Gauge('session_state_count',
                                   'Number of sessions in the specified state',
                                   **metric_keywords)
    }

    def __init__(self, bgpPeering=None, protocol=None):
        self.bgpPeering = bgpPeering
        self.protocol = protocol

        self.connectRetryCounter = 0
        self.connectRetryTime = 30
        self.connectRetryTimer = FSM.BGPTimer(self.connectRetryTimeEvent)
        self.holdTime = 3 * 60
        self.holdTimer = FSM.BGPTimer(self.holdTimeEvent)
        self.keepAliveTime = self.holdTime / 3
        self.keepAliveTimer = FSM.BGPTimer(self.keepAliveEvent)

        self.allowAutomaticStart = True
        self.allowAutomaticStop = False
        self.delayOpen = False
        self.delayOpenTime = 30
        self.delayOpenTimer = FSM.BGPTimer(self.delayOpenEvent)

        self.dampPeerOscillations = True
        self.idleHoldTime = 30
        self.idleHoldTimer = FSM.BGPTimer(self.idleHoldTimeEvent)

        self.metric_labels = {
            'state': stateDescr[self.state],
            'local_asn': None,
            'local_ip': None,
            'remote_ip': None,
            'side': None
        }
        if self.bgpPeering:
            self.metric_labels['local_asn'] = self.bgpPeering.myASN

        self.initial_idle_state = True

    def log(self, msg, lvl=logging.DEBUG):
        s = "bgp.FSM@{}".format(hex(id(self)))
        if self.protocol is not None:
            s += " peer {}".format(self.protocol.peerAddrStr())
        elif self.bgpPeering is not None:
            s += " peer {}".format(self.bgpPeering.peerAddr)
        _log(msg, lvl, s)

    def __setattr__(self, name, value):
        if name == 'state' and value != getattr(self, name):
            self.log("State is now: %s" % stateDescr[value], logging.INFO)
            self.__update_metrics(value)
        super(FSM, self).__setattr__(name, value)

    def __update_metrics(self, new_state):
        if self.metric_labels['local_ip'] and self.metric_labels['remote_ip']:
                if not self.initial_idle_state:
                    self.metrics['bgp_session_state_count'].labels(**self.metric_labels).dec()
                else:
                    self.initial_idle_state = False
                self.metric_labels['state'] = stateDescr[new_state]
                self.metrics['bgp_session_state_count'].labels(**self.metric_labels).inc()

    def manualStart(self):
        """
        Should be called when a BGP ManualStart event (event 1) is requested.
        Note that a protocol instance does not yet exist at this point.
        """

        if self.state == ST_IDLE:
            self.connectRetryCounter = 0
            self.connectRetryTimer.reset(self.connectRetryTime)
            if self.bgpPeering is not None:
                # Create outbound connection
                if self.bgpPeering.connect():
                    self.state = ST_CONNECT

    def manualStop(self):
        """Should be called when a BGP ManualStop event (event 2) is requested."""

        if self.state != ST_IDLE:
            self.protocol.sendNotification(ERR_CEASE, 0)
            # Stop all timers
            for timer in (self.connectRetryTimer, self.holdTimer, self.keepAliveTimer,
                          self.delayOpenTimer, self.idleHoldTimer):
                timer.cancel()
            if self.bgpPeering is not None: self.bgpPeering.releaseResources(self.protocol)
            self._closeConnection()
            self.connectRetryCounter = 0
            self.state = ST_IDLE
            raise NotificationSent(self.protocol, ERR_CEASE, 0)

    def automaticStart(self, idleHold=False):
        """
        Should be called when a BGP Automatic Start event (event 3) is requested.
        Returns True or False to indicate BGPPeering whether a connection attempt
        has been initiated.
        """

        if self.state == ST_IDLE:
            if idleHold:
                self.idleHoldTimer.reset(self.idleHoldTime)
                return False
            else:
                self.connectRetryCounter = 0
                self.connectRetryTimer.reset(self.connectRetryTime)
                # Create outbound connection if possible
                if self.bgpPeering is not None and self.bgpPeering.connect():
                    self.state = ST_CONNECT
                    return True
                else:
                    return False

    def _startPassive(self):
        if self.state == ST_IDLE:
            if self.bgpPeering is not None:
                self.bgpPeering.completeInit(self.protocol)
            self.connectRetryCounter = 0
            self.connectRetryTimer.reset(self.connectRetryTime)
            self.state = ST_ACTIVE

    def manualStartPassive(self):
        """
        Should be called when a BGP Manual Start with Passive TCP Establishment
        (event 4) is requested.
        """

        self._startPassive()

    def automaticStartPassive(self):
        """
        Should be called when a BGP Automatic Start with Passive TCP Establishment
        (event 5) is requested.
        """

        self._startPassive()

    def tcpConnectionValid(self):
        """
        Should be called when a TCP connection request has been received.
        (event 14)
        Returns True if the connection should be accepted, False if not.
        """

        return (self.state != ST_IDLE)

    def connectionMade(self):
        """Should be called when a TCP connection has successfully been
        established with the peer. (events 16, 17)
        """

        if self.state in (ST_CONNECT, ST_ACTIVE):
            # State Connect, Event 16 or 17
            if self.delayOpen:
                self.connectRetryTimer.cancel()
                self.delayOpenTimer.reset(self.delayOpenTime)
            else:
                self.connectRetryTimer.cancel()
                if self.bgpPeering: self.bgpPeering.completeInit(self.protocol)
                self.protocol.sendOpen()
                self.holdTimer.reset(self.largeHoldTime)
                self.state = ST_OPENSENT

    def connectionFailed(self):
        """Should be called when the associated TCP connection failed, or
        was lost. (event 18)"""

        if self.state == ST_CONNECT:
            # State Connect, event 18
            if self.delayOpenTimer.active():
                 self.connectRetryTimer.reset(self.connectRetryTime)
                 self.delayOpenTimer.cancel()
                 self.state = ST_ACTIVE
            else:
                self.connectRetryTimer.cancel()
                self._closeConnection()
                if self.bgpPeering: self.bgpPeering.releaseResources(self.protocol)
                self.state = ST_IDLE
        elif self.state == ST_ACTIVE:
            # State Active, event 18
            self.connectRetryTimer.reset(self.connectRetryTime)
            self.delayOpenTimer.cancel()
            if self.bgpPeering: self.bgpPeering.releaseResources(self.protocol)
            self.connectRetryCounter += 1
            # TODO: osc damping
            self.state = ST_IDLE
        elif self.state == ST_OPENSENT:
            # State OpenSent, event 18
            if self.bgpPeering: self.bgpPeering.releaseResources(self.protocol)
            self._closeConnection()
            self.connectRetryTimer.reset(self.connectRetryTime)
            self.state = ST_ACTIVE
        elif self.state in (ST_OPENCONFIRM, ST_ESTABLISHED):
            self._errorClose()


    def openReceived(self):
        """Should be called when a BGP Open message was received from
        the peer. (events 19, 20)
        """

        if self.state in (ST_CONNECT, ST_ACTIVE):
            if self.delayOpenTimer.active():
                # State Connect, event 20
                self.connectRetryTimer.cancel()
                if self.bgpPeering: self.bgpPeering.completeInit(self.protocol)
                self.delayOpenTimer.cancel()
                self.protocol.sendOpen()
                self.protocol.sendKeepAlive()
                if self.holdTime != 0:
                    self.keepAliveTimer.reset(self.keepAliveTime)
                    self.holdTimer.reset(self.holdTime)
                else:    # holdTime == 0
                    self.keepAliveTimer.cancel()
                    self.holdTimer.cancel()

                self.state = ST_OPENCONFIRM
            else:
                # State Connect, event 19
                self._errorClose()

        elif self.state == ST_OPENSENT:
            if not self.delayOpen:
                # State OpenSent, event 19
                self.delayOpenTimer.cancel()
                self.connectRetryTimer.cancel()
                self.protocol.sendKeepAlive()
                if self.holdTime > 0:
                    self.keepAliveTimer.reset(self.keepAliveTime)
                    self.holdTimer.reset(self.holdTime)
                self.state = ST_OPENCONFIRM
            else:
                # State OpenSent, event 20
                self.protocol.sendNotification(ERR_FSM, 0)
                self._errorClose()
                raise NotificationSent(self.protocol, ERR_FSM, 0)

        elif self.state == ST_OPENCONFIRM:
            if not self.delayOpen:
                # State OpenConfirm, events 19
                self.log("Running collision detection")

                # Perform collision detection
                self.protocol.collisionDetect()
            else:
                # State OpenConfirm, event 20
                self.protocol.sendNotification(ERR_FSM, 0)
                self._errorClose()
                raise NotificationSent(self.protocol, ERR_FSM, 0)

        elif self.state == ST_ESTABLISHED:
            # State Established, event 19 or 20
            self.protocol.sendNotification(ERR_FSM, 0)
            self._errorClose()
            raise NotificationSent(self.protocol, ERR_FSM, 0)

    def headerError(self, suberror, data=''):
        """
        Should be called when an invalid BGP message header was received.
        (event 21)
        """

        if self.state != ST_IDLE:
            self.protocol.sendNotification(ERR_MSG_HDR, suberror, data)
            # Note: RFC4271 states that we should send ERR_FSM in the
            # Established state, which contradicts earlier statements.
            self._errorClose()
            raise NotificationSent(self.protocol, ERR_MSG_HDR, suberror, data)

    def openMessageError(self, suberror, data=''):
        """
        Should be called when an invalid BGP Open message was received.
        (event 22)
        """

        if self.state != ST_IDLE:
            self.protocol.sendNotification(ERR_MSG_OPEN, suberror, data)
            # Note: RFC4271 states that we should send ERR_FSM in the
            # Established state, which contradicts earlier statements.
            self._errorClose()
            raise NotificationSent(self.protocol, ERR_MSG_OPEN, suberror, data)

    def keepAliveReceived(self):
        """
        Should be called when a BGP KeepAlive packet was received
        from the peer. (event 26)
        """

        if self.state == ST_OPENCONFIRM:
            # State OpenSent, event 26
            self.holdTimer.reset(self.holdTime)
            self.state = ST_ESTABLISHED
            self.protocol.deferred.callback(self.protocol)
        elif self.state == ST_ESTABLISHED:
            # State Established, event 26
            self.holdTimer.reset(self.holdTime)
        elif self.state in (ST_CONNECT, ST_ACTIVE):
            # States Connect, Active, event 26
            self._errorClose()
        elif self.state == ST_OPENSENT:
            # State OpenSent, event 26
            self.protocol.sendNotification(ERR_FSM, 0)
            self._errorClose()
            raise NotificationSent(self.protocol, ERR_FSM, 0)

    def versionError(self):
        """
        Should be called when a BGP Notification Open Version Error
        message was received from the peer. (event 24)
        """

        if self.state in (ST_OPENSENT, ST_OPENCONFIRM):
            # State OpenSent/OpenConfirm, event 24
            self.connectRetryTimer.cancel()
            if self.bgpPeering: self.bgpPeering.releaseResources(self.protocol)
            self._closeConnection()
            self.state = ST_IDLE
        elif self.state in (ST_CONNECT, ST_ACTIVE, ST_ESTABLISHED):
            # State Connect/Active/Established, event 24
            self._errorClose()

    def notificationReceived(self, error, suberror):
        """
        Should be called when a BGP Notification message was
        received from the peer. (events 24, 25)
        """

        if error == ERR_MSG_OPEN and suberror == 1:
            # Event 24
            self.versionError()
        else:
            if self.state != ST_IDLE:
                # State != Idle, events 24, 25
                if self.state == ST_OPENSENT:
                    self.protocol.sendNotification(ERR_FSM, 0)
                self._errorClose()

    def updateReceived(self, update):
        """Called when a valid BGP Update message was received. (event 27)"""

        if self.state == ST_ESTABLISHED:
            # State Established, event 27
            if self.holdTime != 0:
                self.holdTimer.reset(self.holdTime)

            self.bgpPeering.update(update)
        elif self.state in (ST_ACTIVE, ST_CONNECT):
            # States Active, Connect, event 27
            self._errorClose()
        elif self.state in (ST_OPENSENT, ST_OPENCONFIRM):
            # States OpenSent, OpenConfirm, event 27
            self.protocol.sendNotification(ERR_FSM, 0)
            self._errorClose()
            raise NotificationSent(self.protocol, ERR_FSM, 0)

    def updateError(self, suberror, data=''):
        """Called when an invalid BGP Update message was received. (event 28)"""

        if self.state == ST_ESTABLISHED:
            # State Established, event 28
            self.protocol.sendNotification(ERR_MSG_UPDATE, suberror, data)
            self._errorClose()
            raise NotificationSent(self.protocol, ERR_MSG_UPDATE, suberror, data)
        elif self.state in (ST_ACTIVE, ST_CONNECT):
            # States Active, Connect, event 28
            self._errorClose()
        elif self.state in (ST_OPENSENT, ST_OPENCONFIRM):
            # States OpenSent, OpenConfirm, event 28
            self.protocol.sendNotification(ERR_FSM, 0)
            self._errorClose()
            raise NotificationSent(self.protocol, ERR_FSM, 0)

    def openCollisionDump(self):
        """
        Called when the collision detection algorithm determined
        that the associated connection should be dumped.
        (event 23)
        """

        self.log("Collided, closing")

        if self.state == ST_IDLE:
            return
        elif self.state in (ST_OPENSENT, ST_OPENCONFIRM, ST_ESTABLISHED):
            self.protocol.sendNotification(ERR_CEASE, 0)

        self._errorClose()
        raise NotificationSent(self.protocol, ERR_CEASE, 0)

    def delayOpenEvent(self):
        """Called when the DelayOpenTimer expires. (event 12)"""

        assert(self.delayOpen)

        self.log("Delay Open event")

        if self.state == ST_CONNECT:
            # State Connect, event 12
            self.protocol.sendOpen()
            self.holdTimer.reset(self.largeHoldTime)
            self.state = ST_OPENSENT
        elif self.state == ST_ACTIVE:
            # State Active, event 12
            self.connectRetryTimer.cancel()
            self.delayOpenTimer.cancel()
            if self.bgpPeering: self.bgpPeering.completeInit(self.protocol)
            self.protocol.sendOpen()
            self.holdTimer.reset(self.largeHoldTime)
            self.state = ST_OPENSENT
        elif self.state != ST_IDLE:
            # State OpenSent, OpenConfirm, Established, event 12
            self.protocol.sendNotification(ERR_FSM, 0)
            self._errorClose()
            raise NotificationSent(self.protocol, ERR_FSM, 0)

    def keepAliveEvent(self):
        """Called when the KeepAliveTimer expires. (event 11)"""

        if self.state in (ST_OPENCONFIRM, ST_ESTABLISHED):
            # State OpenConfirm, Established, event 11
            self.protocol.sendKeepAlive()
            if self.holdTime > 0:
                self.keepAliveTimer.reset(self.keepAliveTime)
        elif self.state in (ST_CONNECT, ST_ACTIVE):
            self._errorClose()
        elif self.state == ST_OPENSENT:
            self.protocol.sendNotification(ERR_FSM, 0)
            self._errorClose()
            raise NotificationSent(self.protocol, ERR_FSM, 0)

    def holdTimeEvent(self):
        """Called when the HoldTimer expires. (event 10)"""

        if self.state in (ST_OPENSENT, ST_OPENCONFIRM, ST_ESTABLISHED):
            # States OpenSent, OpenConfirm, Established, event 10
            self.protocol.sendNotification(ERR_HOLD_TIMER_EXPIRED, 0)
            self._errorClose()
            # TODO: peer osc damping
        elif self.state in (ST_CONNECT, ST_ACTIVE):
            self._errorClose()

    def connectRetryTimeEvent(self):
        """Called when the ConnectRetryTimer expires. (event 9)"""

        if self.state == ST_CONNECT:
            # State Connect, event 9
            self._closeConnection()
            self.connectRetryTimer.reset(self.connectRetryTime)
            self.delayOpenTimer.cancel()
            # Initiate TCP connection
            if self.bgpPeering: self.bgpPeering.connectRetryEvent(self.protocol)
        elif self.state == ST_ACTIVE:
            # State Active, event 9
            self.connectRetryTimer.reset(self.connectRetryTime)
            # Initiate TCP connection
            if self.bgpPeering: self.bgpPeering.connectRetryEvent(self.protocol)
            self.state = ST_CONNECT
        elif self.state != ST_IDLE:
            # State OpenSent, OpenConfirm, Established, event 12
            self.protocol.sendNotification(ERR_FSM, 0)
            self._errorClose()
            raise NotificationSent(self.protocol, ERR_FSM, 0)

    def idleHoldTimeEvent(self):
        """Called when the IdleHoldTimer expires. (event 13)"""

        if self.state == ST_IDLE:
            if self.bgpPeering: self.bgpPeering.automaticStart(idleHold=False)
        else:
            fsmError = False
            if self.state in (ST_OPENSENT, ST_OPENCONFIRM, ST_ESTABLISHED):
                fsmError = True
                self.protocol.sendNotification(ERR_FSM, 0)
            self._errorClose()
            if fsmError:
                raise NotificationSent(self.protocol, ERR_FSM, 0)

    def updateSent(self):
        """Called by the protocol instance when it just sent an Update message."""

        if self.holdTime > 0:
            self.keepAliveTimer.reset(self.keepAliveTime)

    def _errorClose(self):
        """Internal method that closes a connection and returns the state
        to IDLE.
        """

        # Stop the timers
        for timer in (self.connectRetryTimer, self.delayOpenTimer, self.holdTimer,
            self.keepAliveTimer):
            timer.cancel()

        # Release BGP resources (routes, etc)
        if self.bgpPeering: self.bgpPeering.releaseResources(self.protocol)

        self._closeConnection()

        self.connectRetryCounter += 1
        self.state = ST_IDLE

    def _closeConnection(self):
        """Internal method that close the connection if a valid BGP protocol
        instance exists.
        """

        if self.protocol is not None:
            self.protocol.closeConnection()
        # Remove from connections list
        if self.bgpPeering: self.bgpPeering.connectionClosed(self.protocol)
